File: src/lib/proxyUtils.ts
import { URL } from 'url';
import * as HttpsProxyAgentModule from 'https-proxy-agent';
import * as HttpProxyAgentModule from 'http-proxy-agent';

const HttpsProxyAgent = (HttpsProxyAgentModule as any).HttpsProxyAgent || (HttpsProxyAgentModule as any).default;
const HttpProxyAgent = (HttpProxyAgentModule as any).HttpProxyAgent || (HttpProxyAgentModule as any).default;

/**
 * Check if a URL should bypass proxy based on NO_PROXY environment variable
 * @param url - The URL to check
 * @returns true if the URL should bypass proxy, false otherwise
 */
export function shouldBypassProxy(url: string): boolean {
  const noProxy = process.env.NO_PROXY || process.env.no_proxy || '';
  if (!noProxy) {
    return true;
  }

  try {
    const urlObj = new URL(url);
    const hostname = urlObj.hostname;

    // Parse NO_PROXY entries (comma-separated)
    const entries = noProxy.split(',').map((entry) => entry.trim());

    for (const entry of entries) {
      if (!entry) continue;

      // Support wildcards and exact matches
      if (entry === '*') {
        return true;
      }

      // Handle *.example.com pattern
      if (entry.startsWith('*.')) {
        const domain = entry.substring(2);
        // Only match subdomains, not the domain itself
        if (hostname.endsWith('.' + domain)) {
          return true;
        }
      }

      // Handle exact domain match or .domain.com pattern
      if (entry.startsWith('.')) {
        if (hostname.endsWith(entry) || hostname === entry.substring(1)) {
          return true;
        }
      }

      // Handle exact hostname match
      if (hostname === entry) {
        return true;
      }
    }

    return false;
  } catch (_error) {
    return false;
  }
}

/**
 * Get appropriate proxy agent for a URL based on environment variables
 * @param url - The URL to create agent for
 * @returns RequestInit with appropriate agent, or empty object if no proxy needed
 */
export function getProxyAgent(url: string): { agent?: any; } {
  if (shouldBypassProxy(url)) {
    return {};
  }

  try {
    const urlObj = new URL(url);
    const protocol = urlObj.protocol;

    if (protocol === 'https:') {
      const httpsProxy = process.env.HTTPS_PROXY || process.env.https_proxy;
      if (httpsProxy) {
        return { agent: new HttpsProxyAgent(httpsProxy) };
      }
    } else if (protocol === 'http:') {
      const httpProxy = process.env.HTTP_PROXY || process.env.http_proxy;
      if (httpProxy) {
        return { agent: new HttpProxyAgent(httpProxy) };
      }
    }
  } catch (_error) {
    // If URL parsing fails, return empty object
  }

  return {};
}

/**
 * Create fetch options with proxy support
 * @param url - The URL to fetch
 * @param options - Base fetch options
 * @returns Merged fetch options with proxy agent if applicable
 */
export function withProxySupport(url: string, options: RequestInit = {}): RequestInit {
  const proxyOptions = getProxyAgent(url);
  return { ...options, ...proxyOptions };
}

File: internal/utils/proxy_utils.go
package utils

import (
	"net/netip"
	"net/url"
	"strings"
)

// ShouldBypassProxy checks if the target URL should bypass proxy based on no_proxy settings
func ShouldBypassProxy(targetURL string, noProxyEntries []string) bool {
	if len(noProxyEntries) == 0 {
		return false
	}

	u, err := url.Parse(targetURL)
	if err != nil {
		return false
	}

	hostname := u.Hostname()

	for _, entry := range noProxyEntries {
		entry = strings.TrimSpace(entry)
		if entry == "" {
			continue
		}

		// Wildcard matches everything
		if entry == "*" {
			return true
		}

		// CIDR matching (e.g., 10.0.0.0/8)
		if strings.Contains(entry, "/") {
			if matchCIDR(hostname, entry) {
				return true
			}
			continue
		}

		// Wildcard domain (*.example.com)
		if strings.HasPrefix(entry, "*.") {
			domain := entry[2:]
			if strings.HasSuffix(hostname, "."+domain) || hostname == domain {
				return true
			}
			continue
		}

		// Domain suffix (.example.com matches subdomains)
		if strings.HasPrefix(entry, ".") {
			if strings.HasSuffix(hostname, entry) || hostname == entry[1:] {
				return true
			}
			continue
		}

		// Exact match
		if hostname == entry {
			return true
		}
	}

	return false
}

func matchCIDR(ip string, cidr string) bool {
	addr, err := netip.ParseAddr(ip)
	if err != nil {
		return false
	}

	prefix, err := netip.ParsePrefix(cidr)
	if err != nil {
		return false
	}

	return prefix.Contains(addr)
}

// ParseNoProxy parses NO_PROXY environment variable into a list of entries
func ParseNoProxy(noProxy string) []string {
	if noProxy == "" {
		return nil
	}

	entries := strings.Split(noProxy, ",")
	result := make([]string, 0, len(entries))
	for _, entry := range entries {
		entry = strings.TrimSpace(entry)
		if entry != "" {
			result = append(result, entry)
		}
	}
	return result
}